System/OsString/Common.hs

```diff
@@ -313,3 +414,571 @@ toChar (WindowsChar w) = chr $ fromIntegral w
+--
+-- @since 1.4.200.0
+last :: HasCallStack => PLATFORM_STRING -> PLATFORM_WORD
+last = coerce BS.last
+
+-- | /O(n)/ Extract the elements after the head of a OsString, which must be non-empty.
+-- An exception will be thrown in the case of an empty OsString.
+--
+-- This is a partial function, consider using 'uncons' instead.
+--
+-- @since 1.4.200.0
+tail :: HasCallStack => PLATFORM_STRING -> PLATFORM_STRING
+tail = coerce BS.tail
+
+-- | /O(n)/ Extract the 'head' and 'tail' of a OsString, returning 'Nothing'
+-- if it is empty.
+--
+-- @since 1.4.200.0
+uncons :: PLATFORM_STRING -> Maybe (PLATFORM_WORD, PLATFORM_STRING)
+uncons = coerce BS.uncons
+
+-- | /O(1)/ Extract the first element of a OsString, which must be non-empty.
+-- An exception will be thrown in the case of an empty OsString.
+--
+-- This is a partial function, consider using 'uncons' instead.
+--
+-- @since 1.4.200.0
+head :: HasCallStack => PLATFORM_STRING -> PLATFORM_WORD
+head = coerce BS.head
+
+-- | /O(n)/ Return all the elements of a 'OsString' except the last one.
+-- An exception will be thrown in the case of an empty OsString.
+--
+-- This is a partial function, consider using 'unsnoc' instead.
+--
+-- @since 1.4.200.0
+init :: HasCallStack => PLATFORM_STRING -> PLATFORM_STRING
+init = coerce BS.init
+
+-- | /O(n)/ Extract the 'init' and 'last' of a OsString, returning 'Nothing'
+-- if it is empty.
+--
+-- @since 1.4.200.0
+unsnoc :: PLATFORM_STRING -> Maybe (PLATFORM_STRING, PLATFORM_WORD)
+unsnoc = coerce BS.unsnoc
+
+-- | /O(1)/. The empty 'OsString'.
+--
+-- @since 1.4.200.0
+null :: PLATFORM_STRING -> Bool
+null = coerce BS.null
+
+-- | /O(1)/ The length of a 'OsString'.
+--
+-- @since 1.4.200.0
+length :: PLATFORM_STRING -> Int
+length = coerce BS.length
+
+-- | /O(n)/ 'map' @f xs@ is the OsString obtained by applying @f@ to each
+-- element of @xs@.
+--
+-- @since 1.4.200.0
+map :: (PLATFORM_WORD -> PLATFORM_WORD) -> PLATFORM_STRING -> PLATFORM_STRING
+map = coerce BS.map
+
+-- | /O(n)/ 'reverse' @xs@ efficiently returns the elements of @xs@ in reverse order.
+--
+-- @since 1.4.200.0
+reverse :: PLATFORM_STRING -> PLATFORM_STRING
+reverse = coerce BS.reverse
+
+-- | /O(n)/ The 'intercalate' function takes a 'OsString' and a list of
+-- 'OsString's and concatenates the list after interspersing the first
+-- argument between each element of the list.
+--
+-- @since 1.4.200.0
+intercalate :: PLATFORM_STRING -> [PLATFORM_STRING] -> PLATFORM_STRING
+intercalate = coerce BS.intercalate
+
+-- | 'foldl', applied to a binary operator, a starting value (typically
+-- the left-identity of the operator), and a OsString, reduces the
+-- OsString using the binary operator, from left to right.
+--
+-- @since 1.4.200.0
+foldl :: (a -> PLATFORM_WORD -> a) -> a -> PLATFORM_STRING -> a
+#ifdef WINDOWS
+foldl f a (WindowsString s) = BS16.foldl (coerce f) a s
+#else
+foldl f a (PosixString s) = BS8.foldl (coerce f) a s
+#endif
```

Comment:
Humm. `foldl = coerce BS.foldl` doesn't work because the type at which `BS.foldl` is applied is ambiguous. Non-CPP options include:
* `foldl f a s = BS.foldl (coerce f) a (coerce s)`
* `foldl = coerce (BS.foldl @a)` (with `ScopedTypeVariables`)

Arguably the latter is slightly better as it keeps me from wondering if an `INLINE` pragma on `BS.foldl` might fire before eta-reduction happens.